zipline/assets/assets.py

```diff
@@ -0,0 +1,535 @@
+        so we can find a match by inserting an underscore.
+        """
+        symbol = symbol.upper()
+        as_of_date = normalize_date(as_of_date)
+
+        if not fuzzy:
+            try:
+                return self.lookup_symbol_resolve_multiple(symbol, as_of_date)
+            except SymbolNotFound:
+                return None
+        else:
+            try:
+                return self.fuzzy_match[(symbol, fuzzy, as_of_date)]
+            except KeyError:
+                # if symbol is CMCSA and fuzzy is '_', then
+                # try CMCSA, then CMCS_A, then CMC_SA, etc.
+                for fuzzy_symbol in chain(
+                        (symbol,),
+                        (symbol[:i] + fuzzy + symbol[i:]
+                         for i in range(len(symbol) - 1, 0, -1))):
+
+                    infos = self.sym_cache.get(fuzzy_symbol)
+                    if infos:
+                        info, date_match = self._lookup_symbol_in_infos(
+                            infos,
+                            as_of_date,
+                        )
+
+                        if info is not None and date_match:
+                            self.fuzzy_match[(symbol, fuzzy, as_of_date)] = \
+                                info
+                            return info
+                else:
+                    self.fuzzy_match[(symbol, fuzzy, as_of_date)] = None
+
+    def populate_cache(self, force=False):
+        """
+        Populates the asset cache with all values in the assets
+        collection.
+        """
+        if not force and any(c for c in self.shared_caches.items()
+                             if c is not None):
+            return
+
+        # Wipe caches before repopulating
+        self.cache = {}
+        self.sym_cache = {}
+        self.fuzzy_match = {}
+
+        counter = 0
+        for identifier in self.metadata:
+            row = self.metadata.retrieve_metadata(identifier=identifier)
+            self.spawn_asset(identifier=identifier, **row)
+            counter += 1
+
+        self.shared_caches.update(
+            {'by_sid': self.cache,
+             'by_symbol': self.sym_cache,
+             'fuzzy_match': {}}
+        )
+
+    def spawn_asset(self, identifier, **kwargs):
+
+        # Check if the identifier is an int
+        if isinstance(identifier, int) and ('sid' not in kwargs.keys()):
+            kwargs['sid'] = identifier
+
+        # Make sure that the sid exists in the kwargs
+        if 'sid' not in kwargs.keys():
+            kwargs['sid'] = self._assign_sid(identifier)
+
+        # If the identifier coming in was a string and there is no defined
+        # symbol yet, set the symbol to the incoming sid
+        if isinstance(identifier, str) and ('symbol' not in kwargs.keys()):
+            kwargs['symbol'] = identifier
+
+        # If the file_name is in the kwargs, it may be the symbol
+        if 'file_name' in kwargs.keys():
```

Comment:
To avoid mentioning every line, the note about checking `.keys()` applies to the rest of this function.